### DISCUSSION TITLE
Inner product index but cosine distance matching?

### DISCUSSION THREAD

**Discussion by helloivanco** [kaarthik108/snowChat#15] [Q&A]
Hi,

I have a question on the supabase sql. It appears the index is created for Inner product (vector_ip_ops) but the v_match_documents is matching based on `<=>` the cosine distance? Should the match function change to match inner product or the index be created for `vector_cosine_ops`? 

```
CREATE INDEX ON documents USING hnsw (embedding vector_ip_ops);
CREATE FUNCTION v_match_documents (
    query_embedding vector (1536),
    filter jsonb default '{}'
) RETURNS table (
    id uuid,
    content text,
    metadata jsonb,
    similarity float
) language plpgsql as $$ #variable_conflict use_column
begin return query
select id,
    content,
    metadata,
    1 - (documents.embedding <=> query_embedding) as similarity
from documents
where metadata @> filter
order by documents.embedding <=> query_embedding;
END;
$$;
```

ref: https://supabase.com/docs/guides/ai/vector-indexes/hnsw-indexes

Thanks,
Ivan

**Comment by kaarthik108** [comment]
Oh right, yes this can be changed to 

` CREATE INDEX ON documents USING hnsw (embedding vector_cosine_ops);
`

Thanks for pointing that 👍

**Reply by helloivanco** [reply]
This seems to be the correct way but from my initial test is resulting in less than optimal embedding matches (https://github.com/orgs/supabase/discussions/13977#discussioncomment-7171942).

The reason why I ran into this issue is I also had the wrong index type and while things appeared to work for a small set of embeddings once the embedding table grew to over 100k records, Supabase started to timeout when retrieving matches.

**Reply by helloivanco** [reply]
**UPDATE**

With index `CREATE INDEX ON documents USING hnsw (embedding vector_cosine_ops);` this updated match function seems to work fine now. 

```
-- Create a function to search for documents
create function match_documents (
  query_embedding vector(1536),
  match_count int DEFAULT 3,
  match_threshold float DEFAULT 0.78,
  filter jsonb DEFAULT '{}'
) returns table (
  id bigint,
  content text,
  metadata jsonb,
  similarity float
)
language plpgsql
as $$
#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    1 - (documents.embedding <=> query_embedding) as similarity
  from documents
  where metadata @> filter and 1 - (documents.embedding <=> query_embedding) > match_threshold
  order by similarity desc
  limit match_count;
end;
$$;
```

**Reply by kaarthik108** [reply]
Thanks @helloivanco I will have a look at this today 💪

**Comment by helloivanco** [comment]
@kaarthik108 - Greg from Supabase has some pointers in this thread https://github.com/orgs/supabase/discussions/13977#discussioncomment-7188584

Hope it all helps.

**Reply by kaarthik108** [reply]
Thanks @helloivanco
